### PR TITLE
ZCS-9067 Adding change password rules.

### PR DESCRIPTION
**Pull request by BhavinBathani**
Adding more strict and early verification rules while user is changing password.
Users are shown the password rules that apply to them.
Immediate feedback is given as a new password is entered, in the form of checkmarks next to the given rules.

### CODE REVIEW

File: WebRoot/h/changepass [hunk 1]
```diff
@@ -80,63 +83,279 @@
...
+            const uppers = [],
+                lowers = [],
+                numbers = [],
+                punctuations = [],
+                invalidChars = [],
+                invalidPuncs = [];
+
+            const chars = passwordString.split('');
+
+            chars.forEach(char => {
```
Comment (silentsakky) [inline WebRoot/h/changepass:264]: not sure all browsers will be supporting arrow functions

File: WebRoot/h/changepass [hunk 2]
```diff
@@ -80,63 +83,279 @@
...
+                const charCode = char.charCodeAt(0);
+                let isInvalid = false;
+
+                if ("${zimbraPasswordAllowedChars}") {
```
Comment (silentsakky) [inline WebRoot/h/changepass:268]: is this ${ used by jsp?

**Comment by silentsakky** [comment]
- [ ] Same flow should be used when enforcing change password after login, check these validations there as well
- [ ] also after doing forget password, we ask user to change password, check if it is using same screen

File: WebRoot/h/changepass [hunk 3]
```diff
@@ -80,63 +83,279 @@
...
+            checkImg.style.display = "inline";
+        }
+        function unCheck(checkImg, closeImg) {
+            closeImg.style.display = "inline";
+            checkImg.style.display = "none";
+        }
+        function resetImg(condition, checkImg, closeImg){
+            condition ? check(checkImg, closeImg) : unCheck(checkImg, closeImg);
+        }
+        function showPassword(x, y, z) {
```
Comment (silentsakky) [inline WebRoot/h/changepass:214]: better names in place of x,y,z

File: WebRoot/h/changepass [hunk 4]
```diff
@@ -80,63 +83,279 @@
...
+                x.type = "text";   
+                y.style.display = "none";
+                z.style.display = "block";
+            } else {
+                x.type = "password";
+                y.style.display = "block";
+                z.style.display = "none";
+            }
+        }
+        function showOldPassword() {
```
Comment (silentsakky) [inline WebRoot/h/changepass:225]: where are we doing match of old password with new password?

File: WebRoot/h/changepass (hunk 2 above)
Comment (BhavinBathani) [reply WebRoot/h/changepass:268]: yes it is

File: WebRoot/h/changepass (hunk 1 above)
Comment (BhavinBathani) [reply WebRoot/h/changepass:264]: Fixed.

**Comment by BhavinBathani** [comment]
> * [ ]  Same flow should be used when enforcing change password after login, check these validations there as well
> * [ ]  also after doing forget password, we ask user to change password, check if it is using same screen

Both these scenarios are not trivial to fix in this particular ticket.
Let me create separate tickets with some engineering inputs as we discussed already.

File: WebRoot/h/changepass (hunk 3 above)
Comment (BhavinBathani) [reply WebRoot/h/changepass:214]: Fixed.

File: WebRoot/h/changepass (hunk 4 above)
Comment (BhavinBathani) [reply WebRoot/h/changepass:237]: Fixed.

**Pull request merged**